File: clients/python/locking.py
"""High-level locking for the Fiducia Python client.

Hand-written (NOT generated): adds the live-mutex-style ergonomics on top of the
generated thin ``FiduciaClient`` in ``fiducia.py``. Two shapes:

    * ``try_lock(key)``  — wait=False. Returns a ``Lock`` if free right now, else
      ``None`` (never blocks).
    * ``lock(key)`` / ``must_lock(key)`` — wait=True. Blocks (polling with a fixed
      interval) until acquired, the budget elapses (``LockTimeout``), or the
      server errors.

Counting semaphores get the same pair (``try_semaphore`` / ``acquire_semaphore``).

The server never holds a request open: ``wait=True`` reserves a FIFO queue slot
and returns immediately, so the *client* owns the wait — hence ``max_wait_ms``,
``retry_interval_ms`` and ``max_retries`` live here.

    from fiducia import FiduciaClient
    from locking import FiduciaLockClient

    c = FiduciaLockClient("https://api.fiducia.cloud")
    lock = c.lock("orders/checkout", ttl_ms=30000)
    try:
        ...  # critical section
    finally:
        lock.release()
"""
import time as _time
import uuid as _uuid

from fiducia import FiduciaClient


class LockTimeout(Exception):
    def __init__(self, keys, waited_ms):
        super().__init__("fiducia: timed out after %dms waiting for %s" % (waited_ms, keys))
        self.keys, self.waited_ms = keys, waited_ms


class Lock:
    """A held lock grant. Call ``release()`` (alias ``unlock()``) when done."""

    def __init__(self, client, keys, holder, fencing_token, lease_expires_ms=None, ttl_ms=60000):
        self._client = client
        self.keys = keys
        self.holder = holder
        self.fencing_token = fencing_token
        self.lease_expires_ms = lease_expires_ms
        self._ttl_ms = ttl_ms

    def renew(self, ttl_ms=None):
        response = self._client.lock_renew(
            self.keys, self.holder, self.fencing_token, ttl_ms or self._ttl_ms
        )
        output = _output(response)
        if not output.get("renewed"):
            raise RuntimeError("fiducia: lock renewal lost fenced authority")
        self.lease_expires_ms = output.get("lease_expires_ms")
        self._ttl_ms = ttl_ms or self._ttl_ms
        return response

    def release(self):
        return self._client.lock_release(self.holder, self.fencing_token)

    unlock = release

    # Context-manager sugar: `with c.lock("k") as lock: ...`
    def __enter__(self):
        return self

    def __exit__(self, *exc):
        try:
            self.release()
        except Exception:
            pass  # best-effort; the lease TTL is the backstop
        return False


class SemaphoreHandle:
    def __init__(self, client, key, holder, fencing_token, lease_expires_ms=None, ttl_ms=60000):
        self._client = client
        self.key = key
        self.holder = holder
        self.fencing_token = fencing_token
        self.lease_expires_ms = lease_expires_ms
        self._ttl_ms = ttl_ms

    def renew(self, ttl_ms=None):
        response = self._client.semaphore_renew(
            self.key, self.holder, self.fencing_token, ttl_ms or self._ttl_ms
        )
        output = _output(response)
        if not output.get("renewed"):
            raise RuntimeError("fiducia: semaphore renewal lost fenced authority")
        self.lease_expires_ms = output.get("lease_expires_ms")
        self._ttl_ms = ttl_ms or self._ttl_ms
        return response

    def release(self):
        return self._client.semaphore_release(self.key, self.holder, self.fencing_token)

    unlock = release

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        try:
            self.release()
        except Exception:
            pass
        return False


def _gen_holder():
    return "fdc-%s" % _uuid.uuid4().hex


def _gen_request_id():
    return "fdc-attempt-%s" % _uuid.uuid4().hex


def _output(resp):
    return ((resp or {}).get("result") or {}).get("output") or {}


def _fencing_token(output, primitive):
    token = output.get("fencing_token")
    if not isinstance(token, int) or isinstance(token, bool) or token <= 0:
        raise RuntimeError("fiducia: acquired %s carried no valid fencing token" % primitive)
    return token


class FiduciaLockClient(FiduciaClient):
    # --- locks ---------------------------------------------------------------

    def try_lock(self, key_or_keys, ttl_ms=60000, holder=None):
        """Take the lock now (wait=False). Returns a Lock or None if held."""
        return self._acquire_lock(_as_list(key_or_keys), False, ttl_ms, holder)

    def lock(self, key_or_keys, ttl_ms=60000, holder=None,
             max_wait_ms=30000, retry_interval_ms=250, max_retries=None):
        """Block until acquired, the budget elapses (LockTimeout), or error."""
        got = self._acquire_lock(_as_list(key_or_keys), True, ttl_ms, holder,
                                 max_wait_ms, retry_interval_ms, max_retries)
        if got is None:
            raise LockTimeout(_as_list(key_or_keys), max_wait_ms)
        return got

    # `must_lock` is an alias of `lock`.
    must_lock = lock

    def with_lock(self, key_or_keys, fn, **opts):
        """Acquire, run ``fn(lock)``, then always release."""
        lock = self.lock(key_or_keys, **opts)
        try:
            return fn(lock)
        finally:
            try:
                lock.release()
            except Exception:
                pass

    def _acquire_lock(self, keys, wait, ttl_ms, holder,
                      max_wait_ms=30000, retry_interval_ms=250, max_retries=None):
        holder = (holder or "").strip() or _gen_holder()
        request_id = _gen_request_id()
        try:
            first = self.lock_acquire_many(
                keys,
                holder=holder,
                request_id=request_id,
                ttl_ms=ttl_ms,
                wait=wait,
                wait_timeout_ms=max_wait_ms if wait else None,
            )
        except Exception:
            self._cancel_lock_wait(keys, holder, request_id)
            raise
        out = _output(first)
        if out.get("acquired"):
            token = _fencing_token(out, "lock")
            if out.get("renewed") is False:
                try:
                    renewed = _output(self.lock_renew(keys, holder, token, ttl_ms))
                    if not renewed.get("renewed"):
                        raise RuntimeError(
                            "fiducia: reacquired lock lost fenced authority during renewal"
                        )
                    return Lock(
                        self, keys, holder, token, renewed.get("lease_expires_ms"), ttl_ms
                    )
                except Exception:
                    self._cancel_lock_wait(keys, holder, request_id)
                    raise
            return Lock(
                self, keys, holder, token, out.get("lease_expires_ms"), ttl_ms
            )
        if not wait:
            return None  # try_lock: held now → fail fast

        deadline = _time.monotonic() + max_wait_ms / 1000.0
        attempts = 0
        acquired = False
        try:
            while max_retries is None or attempts < max_retries:
                remaining = deadline - _time.monotonic()
                if remaining <= 0:
                    break
                delay_ms = min(retry_interval_ms * (2 ** min(attempts, 3)), 2000)
                _time.sleep(min(delay_ms / 1000.0, remaining))
                attempts += 1
                retried = self.lock_acquire_many(
                    keys,
                    holder=holder,
                    request_id=request_id,
                    ttl_ms=ttl_ms,
                    wait=True,
                    wait_timeout_ms=max_wait_ms,
                )
                out = _output(retried)
                if out.get("acquired"):
                    token = _fencing_token(out, "lock")
                    renewed = _output(self.lock_renew(keys, holder, token, ttl_ms))
                    if not renewed.get("renewed"):
                        raise RuntimeError(
                            "fiducia: retry-discovered lock lost fenced authority during renewal"
                        )
                    acquired = True
                    return Lock(
                        self,
                        keys,
                        holder,
                        token,
                        renewed.get("lease_expires_ms"),
                        ttl_ms,
                    )
            return None
        finally:
            if not acquired:
                self._cancel_lock_wait(keys, holder, request_id)

    def _cancel_lock_wait(self, keys, holder, request_id):
        out = _output(self.lock_cancel(keys, holder, request_id=request_id))
        if out.get("acquired"):
            token = _fencing_token(out, "raced lock")
            released = _output(self.lock_release(holder, token))
            if released.get("released") is False:
                raise RuntimeError("fiducia: raced lock could not be released safely")
            return
        if out.get("cancelled") is True:
            return
        raise RuntimeError(
            "fiducia: lock cancellation did not establish safety (%s)"
            % out.get("reason", "invalid_response")
        )

    # --- counting semaphores -------------------------------------------------

    def try_semaphore(self, key, limit, ttl_ms=60000, holder=None):
        """Take a permit now (wait=False). Returns a handle or None if full."""
        return self._acquire_semaphore(key, limit, False, ttl_ms, holder)

    def acquire_semaphore(self, key, limit, ttl_ms=60000, holder=None,
                          max_wait_ms=30000, retry_interval_ms=250, max_retries=None):
        """Block until a permit is free, the budget elapses, or error."""
        got = self._acquire_semaphore(key, limit, True, ttl_ms, holder,
                                      max_wait_ms, retry_interval_ms, max_retries)
        if got is None:
            raise LockTimeout([key], max_wait_ms)
        return got

    def _acquire_semaphore(self, key, limit, wait, ttl_ms, holder,
                           max_wait_ms=30000, retry_interval_ms=250, max_retries=None):
        holder = (holder or "").strip() or _gen_holder()
        request_id = _gen_request_id()
        try:
            first = self.semaphore_acquire(
                key,
                limit,
                holder=holder,
                request_id=request_id,
                ttl_ms=ttl_ms,
                wait=wait,
                wait_timeout_ms=max_wait_ms if wait else None,
            )
        except Exception:
            self._cancel_semaphore_wait(key, holder, request_id)
            raise
        out = _output(first)
        if out.get("acquired"):
            token = _fencing_token(out, "semaphore permit")
            if out.get("renewed") is False:
                try:
                    renewed = _output(self.semaphore_renew(key, holder, token, ttl_ms))
                    if not renewed.get("renewed"):
                        raise RuntimeError(
                            "fiducia: reacquired semaphore permit lost fenced authority during renewal"
                        )
                    return SemaphoreHandle(
                        self, key, holder, token, renewed.get("lease_expires_ms"), ttl_ms
                    )
                except Exception:
                    self._cancel_semaphore_wait(key, holder, request_id)
                    raise
            return SemaphoreHandle(
                self, key, holder, token,
                out.get("lease_expires_ms"), ttl_ms
            )
        if out.get("reason") == "limit_mismatch":
            raise RuntimeError(
                "fiducia: semaphore limit mismatch (requested %s, configured %s)"
                % (limit, out.get("limit"))
            )
        if not wait:
            return None

        deadline = _time.monotonic() + max_wait_ms / 1000.0
        attempts = 0
        acquired = False
        try:
            while max_retries is None or attempts < max_retries:
                remaining = deadline - _time.monotonic()
                if remaining <= 0:
                    break
                delay_ms = min(retry_interval_ms * (2 ** min(attempts, 3)), 2000)
                _time.sleep(min(delay_ms / 1000.0, remaining))
                attempts += 1
                retried = self.semaphore_acquire(
                    key,
                    limit,
                    holder=holder,
                    request_id=request_id,
                    ttl_ms=ttl_ms,
                    wait=True,
                    wait_timeout_ms=max_wait_ms,
                )
                out = _output(retried)
                if out.get("acquired"):
                    token = _fencing_token(out, "semaphore permit")
                    renewed = _output(self.semaphore_renew(key, holder, token, ttl_ms))
                    if not renewed.get("renewed"):
                        raise RuntimeError(
                            "fiducia: retry-discovered semaphore permit lost fenced authority during renewal"
                        )
                    acquired = True
                    return SemaphoreHandle(
                        self,
                        key,
                        holder,
                        token,
                        renewed.get("lease_expires_ms"),
                        ttl_ms,
                    )
                if out.get("reason") == "limit_mismatch":
                    raise RuntimeError(
                        "fiducia: semaphore limit mismatch (requested %s, configured %s)"
                        % (limit, out.get("limit"))
                    )
            return None
        finally:
            if not acquired:
                self._cancel_semaphore_wait(key, holder, request_id)

    def _cancel_semaphore_wait(self, key, holder, request_id):
        out = _output(self.semaphore_cancel(key, holder, request_id=request_id))
        if out.get("acquired"):
            token = _fencing_token(out, "raced semaphore permit")
            released = _output(self.semaphore_release(key, holder, token))
            if released.get("released") is False:
                raise RuntimeError(
                    "fiducia: raced semaphore permit could not be released safely"
                )
            return
        if out.get("cancelled") is True:
            return
        raise RuntimeError(
            "fiducia: semaphore cancellation did not establish safety (%s)"
            % out.get("reason", "invalid_response")
        )


def _as_list(key_or_keys):
    return list(key_or_keys) if isinstance(key_or_keys, (list, tuple)) else [key_or_keys]

File: clients/ts/locking.ts
// High-level locking helpers for the Fiducia TypeScript client.
//
// Hand-written (NOT generated): builds the blocking/try/retry ergonomics on top
// of the generated thin `FiduciaClient` in ./fiducia. Mirrors the live-mutex
// client surface so callers get the two shapes they actually want:
//
//   * tryLock(key)   — wait:false. Returns immediately: a Lock if it was free
//                      right now, or null if it's held (no waiting).
//   * lock(key)      — wait:true. Blocks (polling with backoff) until the lock
//   * mustLock(key)    is acquired, the deadline passes (throws LockTimeoutError),
//                      or the server errors. `mustLock` is an alias of `lock`.
//
// Counting semaphores get the same pair: trySemaphore / acquireSemaphore.
//
// The server never holds a request open: `wait:true` reserves a FIFO queue slot
// and returns immediately, so the *client* owns the wait — that's why retry
// cadence (retryInterval), the total budget (maxWaitTime), and the attempt cap
// (maxRetries) all live here, giving the caller full control.

import { FiduciaClient } from "./fiducia.ts";

type LockClientBase = new (
  ...args: ConstructorParameters<typeof FiduciaClient>
) => Omit<FiduciaClient, "tryLock" | "lock" | "mustLock" | "trySemaphore">;

// Runtime inheritance keeps the complete raw client surface. Omitting only the
// four names replaced by handle-oriented helpers prevents incompatible method
// overrides from erasing type safety for every other operation.
const FiduciaLockClientBase = FiduciaClient as unknown as LockClientBase;

/** Thrown by `lock`/`acquireSemaphore` when the wait budget elapses unacquired. */
export class LockTimeoutError extends Error {
  keys: string[];
  waitedMs: number;

  constructor(keys: string[], waitedMs: number) {
    super(`fiducia: timed out after ${waitedMs}ms waiting for lock on ${keys.join(", ")}`);
    this.name = "LockTimeoutError";
    this.keys = keys;
    this.waitedMs = waitedMs;
  }
}

/** Thrown when a wait is cancelled via an AbortSignal. */
export class LockAbortedError extends Error {
  keys: string[];

  constructor(keys: string[]) {
    super(`fiducia: lock wait aborted for ${keys.join(", ")}`);
    this.name = "LockAbortedError";
    this.keys = keys;
  }
}

export interface LockOptions {
  /** Lease TTL in ms — the lock auto-expires if not released (default 60000). */
  ttl?: number;
  /** Caller identity (and the key for release). Defaults to a generated id. */
  holder?: string;
  /** Total time to keep waiting before giving up, for blocking `lock` (default 30000). */
  maxWaitTime?: number;
  /** Max number of poll attempts while waiting (default unlimited). */
  maxRetries?: number;
  /** Delay between polls while waiting, in ms (default 250). */
  retryInterval?: number;
  /** Cancel an in-flight wait. */
  signal?: AbortSignal;
}

export interface Lock {
  /** The member keys held by this grant (a single-key lock has one). */
  keys: string[];
  /** The holder id that owns the grant (needed to release). */
  holder: string;
  /** Monotonic fencing token — pass it to downstream systems to defeat stale holders. */
  fencingToken: number;
  /** When the lease expires (ms since epoch), if the server reported it. */
  leaseExpiresMs?: number;
  /** Extend this exact fenced grant. Throws if authority has been lost. */
  renew(ttlMs?: number): Promise<any>;
  /** Release the whole grant. Idempotent-ish; safe to call once. */
  unlock(): Promise<any>;
  /** Alias of `unlock`. */
  release(): Promise<any>;
}

export interface SemaphoreHandle {
  key: string;
  holder: string;
  fencingToken: number;
  leaseExpiresMs?: number;
  /** Extend this exact fenced permit. Throws if authority has been lost. */
  renew(ttlMs?: number): Promise<any>;
  unlock(): Promise<any>;
  release(): Promise<any>;
}

const DEFAULT_TTL = 60_000;
const DEFAULT_MAX_WAIT = 30_000;
const DEFAULT_RETRY_INTERVAL = 250;

function genId(): string {
  const cryptoApi = globalThis.crypto;
  if (cryptoApi?.randomUUID) return `fdc-${cryptoApi.randomUUID()}`;
  if (cryptoApi?.getRandomValues) {
    const bytes = new Uint8Array(16);
    cryptoApi.getRandomValues(bytes);
    return `fdc-${Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("")}`;
  }
  throw new Error("fiducia: secure randomness is unavailable; provide an explicit nonempty holder");
}

function fencingToken(output: any, primitive: string): number {
  const token = output?.fencing_token;
  if (!Number.isSafeInteger(token) || token <= 0) {
    throw new Error(`fiducia: acquired ${primitive} carried no valid fencing token`);
  }
  return token;
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    const t = setTimeout(resolve, ms);
    if (signal) {
      const onAbort = () => {
        clearTimeout(t);
        reject(new Error("aborted"));
      };
      if (signal.aborted) onAbort();
      else signal.addEventListener("abort", onAbort, { once: true });
    }
  });
}

/**
 * `FiduciaClient` plus high-level lock/semaphore acquisition. Use this in place
 * of `FiduciaClient` when you want `lock`/`tryLock` rather than the raw calls.
 */
export class FiduciaLockClient extends FiduciaLockClientBase {
  // --- locks ---------------------------------------------------------------

  /** Try to take the lock right now (wait:false). Returns null if it's held. */
  async tryLock(keyOrKeys: string | string[], opts: LockOptions = {}): Promise<Lock | null> {
    return this.acquireLock(normalize(keyOrKeys), false, opts);
  }

  /**
   * Block until the lock is acquired, the wait budget elapses
   * (`LockTimeoutError`), or the server errors. `wait:true`.
   */
  async lock(keyOrKeys: string | string[], opts: LockOptions = {}): Promise<Lock> {
    const lock = await this.acquireLock(normalize(keyOrKeys), true, opts);
    // acquireLock(wait=true) only returns null on a logic error; treat as timeout.
    if (!lock) throw new LockTimeoutError(normalize(keyOrKeys), opts.maxWaitTime ?? DEFAULT_MAX_WAIT);
    return lock;
  }

  /** Alias of `lock` — blocks until acquired (or throws). */
  async mustLock(keyOrKeys: string | string[], opts: LockOptions = {}): Promise<Lock> {
    return this.lock(keyOrKeys, opts);
  }

  /** Acquire, run `fn`, and always release (even if `fn` throws). */
  async withLock<T>(
    keyOrKeys: string | string[],
    opts: LockOptions,
    fn: (lock: Lock) => Promise<T> | T,
  ): Promise<T> {
    const lock = await this.lock(keyOrKeys, opts);
    try {
      return await fn(lock);
    } finally {
      try {
        await lock.unlock();
      } catch {
        /* best-effort release; the lease TTL is the backstop */
      }
    }
  }

  private async acquireLock(
    keys: string[],
    wait: boolean,
    opts: LockOptions,
  ): Promise<Lock | null> {
    const holder = opts.holder?.trim() || genId();
    const requestId = genId();
    const ttl = opts.ttl ?? DEFAULT_TTL;

    const maxWait = opts.maxWaitTime ?? DEFAULT_MAX_WAIT;
    let first: any;
    try {
      first = await this.lockAcquireMany({
        keys,
        holder,
        requestId,
        ttlMs: ttl,
        wait,
        waitTimeoutMs: wait ? maxWait : undefined,
      });
    } catch (error) {
      // The request may have committed even when its response was lost. Cancel
      // the identity and release any raced grant before surfacing ambiguity.
      await this.cancelLockWait(keys, holder, requestId);
      throw error;
    }
    const out = first?.result?.output ?? {};
    if (out.acquired) {
      const token = fencingToken(out, "lock");
      if (out.renewed === false) {
        try {
          const renewed = await this.lockRenew(keys, holder, token, ttl);
          const renewedOut = renewed?.result?.output ?? {};
          if (!renewedOut.renewed) {
            throw new Error("fiducia: reacquired lock lost fenced authority during renewal");
          }
          return this.lockHandle(keys, holder, token, renewedOut.lease_expires_ms, ttl);
        } catch (error) {
          await this.cancelLockWait(keys, holder, requestId);
          throw error;
        }
      }
      return this.lockHandle(keys, holder, token, out.lease_expires_ms, ttl);
    }
    if (!wait) return null; // tryLock: held right now → fail fast

    // Re-submit the exact queued identity with bounded backoff. A committed
    // retry advances expiry/promotion in the Raft state machine, while queue
    // dedup preserves the original FIFO position.
    const deadline = Date.now() + maxWait;
    const interval = opts.retryInterval ?? DEFAULT_RETRY_INTERVAL;
    const maxRetries = opts.maxRetries ?? Infinity;
    let acquired = false;
    try {
      for (let attempt = 0; attempt < maxRetries; attempt++) {
        const remaining = deadline - Date.now();
        if (remaining <= 0) break;
        try {
          const delay = Math.min(interval * (2 ** Math.min(attempt, 3)), 2_000, remaining);
          await sleep(delay, opts.signal);
        } catch {
          throw new LockAbortedError(keys);
        }
        const retried = await this.lockAcquireMany({
          keys,
          holder,
          requestId,
          ttlMs: ttl,
          wait: true,
          waitTimeoutMs: maxWait,
        });
        const retryOut = retried?.result?.output ?? {};
        if (retryOut.acquired) {
          const token = fencingToken(retryOut, "lock");
          const renewed = await this.lockRenew(keys, holder, token, ttl);
          const renewedOut = renewed?.result?.output ?? {};
          if (!renewedOut.renewed) {
            throw new Error("fiducia: retry-discovered lock lost fenced authority during renewal");
          }
          acquired = true;
          return this.lockHandle(
            keys,
            holder,
            token,
            renewedOut.lease_expires_ms,
            ttl,
          );
        }
      }
      throw new LockTimeoutError(keys, maxWait);
    } finally {
      if (!acquired) await this.cancelLockWait(keys, holder, requestId);
    }
  }

  private lockHandle(
    keys: string[],
    holder: string,
    fencingToken: number,
    leaseExpiresMs?: number,
    defaultTtlMs: number = DEFAULT_TTL,
  ): Lock {
    let currentTtlMs = defaultTtlMs;
    const handle: Lock = {
      keys,
      holder,
      fencingToken,
      leaseExpiresMs,
      renew: async (ttlMs = currentTtlMs) => {
        const response = await this.lockRenew(keys, holder, fencingToken, ttlMs);
        const output = response?.result?.output ?? {};
        if (!output.renewed) throw new Error("fiducia: lock renewal lost fenced authority");
        handle.leaseExpiresMs = output.lease_expires_ms;
        currentTtlMs = ttlMs;
        return response;
      },
      unlock: () => this.lockRelease(keys[0], { holder, fencingToken }),
      release: () => this.lockRelease(keys[0], { holder, fencingToken }),
    };
    return handle;
  }

  private async cancelLockWait(keys: string[], holder: string, requestId: string): Promise<void> {
    const response = await this.lockCancel(keys, holder, { requestId });
    const output = response?.result?.output ?? {};
    if (output.acquired && output.fencing_token != null) {
      const released = await this.lockRelease(keys[0], {
        holder,
        fencingToken: fencingToken(output, "raced lock"),
      });
      if (released?.result?.output?.released === false) {
        throw new Error("fiducia: raced lock could not be released safely");
      }
      return;
    }
    if (output.cancelled === true) return;
    throw new Error(
      `fiducia: lock cancellation did not establish safety (${String(output.reason ?? "invalid_response")})`,
    );
  }

  // --- counting semaphores -------------------------------------------------

  /** Take a permit right now (wait:false). Returns null if at capacity. */
  async trySemaphore(
    key: string,
    limit: number,
    opts: LockOptions = {},
  ): Promise<SemaphoreHandle | null> {
    return this.acquireSemaphoreInner(key, limit, false, opts);
  }

  /** Block until a permit is free, the budget elapses, or the server errors. */
  async acquireSemaphore(
    key: string,
    limit: number,
    opts: LockOptions = {},
  ): Promise<SemaphoreHandle> {
    const h = await this.acquireSemaphoreInner(key, limit, true, opts);
    if (!h) throw new LockTimeoutError([key], opts.maxWaitTime ?? DEFAULT_MAX_WAIT);
    return h;
  }

  /** Acquire a permit, run `fn`, and always release. */
  async withSemaphore<T>(
    key: string,
    limit: number,
    opts: LockOptions,
    fn: (h: SemaphoreHandle) => Promise<T> | T,
  ): Promise<T> {
    const h = await this.acquireSemaphore(key, limit, opts);
    try {
      return await fn(h);
    } finally {
      try {
        await h.unlock();
      } catch {
        /* best-effort */
      }
    }
  }

  private async acquireSemaphoreInner(
    key: string,
    limit: number,
    wait: boolean,
    opts: LockOptions,
  ): Promise<SemaphoreHandle | null> {
    const holder = opts.holder?.trim() || genId();
    const requestId = genId();
    const ttl = opts.ttl ?? DEFAULT_TTL;

    const maxWait = opts.maxWaitTime ?? DEFAULT_MAX_WAIT;
    let first: any;
    try {
      first = await this.semaphoreAcquire(key, {
        max: limit,
        holder,
        requestId,
        ttlMs: ttl,
        wait,
        waitTimeoutMs: wait ? maxWait : undefined,
      });
    } catch (error) {
      await this.cancelSemaphoreWait(key, holder, requestId);
      throw error;
    }
    const out = first?.result?.output ?? {};
    if (out.acquired) {
      const token = fencingToken(out, "semaphore permit");
      if (out.renewed === false) {
        try {
          const renewed = await this.semaphoreRenew(key, holder, token, ttl);
          const renewedOut = renewed?.result?.output ?? {};
          if (!renewedOut.renewed) {
            throw new Error(
              "fiducia: reacquired semaphore permit lost fenced authority during renewal",
            );
          }
          return this.semaphoreHandle(
            key,
            holder,
            token,
            renewedOut.lease_expires_ms,
            ttl,
          );
        } catch (error) {
          await this.cancelSemaphoreWait(key, holder, requestId);
          throw error;
        }
      }
      return this.semaphoreHandle(
        key,
        holder,
        token,
        out.lease_expires_ms,
        ttl,
      );
    }
    if (out.reason === "limit_mismatch") {
      throw new Error(
        `fiducia: semaphore limit mismatch (requested ${limit}, configured ${String(out.limit)})`,
      );
    }
    if (!wait) return null;

    const deadline = Date.now() + maxWait;
    const interval = opts.retryInterval ?? DEFAULT_RETRY_INTERVAL;
    const maxRetries = opts.maxRetries ?? Infinity;
    let acquired = false;
    try {
      for (let attempt = 0; attempt < maxRetries; attempt++) {
        const remaining = deadline - Date.now();
        if (remaining <= 0) break;
        try {
          const delay = Math.min(interval * (2 ** Math.min(attempt, 3)), 2_000, remaining);
          await sleep(delay, opts.signal);
        } catch {
          throw new LockAbortedError([key]);
        }
        const retried = await this.semaphoreAcquire(key, {
          max: limit,
          holder,
          requestId,
          ttlMs: ttl,
          wait: true,
          waitTimeoutMs: maxWait,
        });
        const retryOut = retried?.result?.output ?? {};
        if (retryOut.acquired) {
          const token = fencingToken(retryOut, "semaphore permit");
          const renewed = await this.semaphoreRenew(key, holder, token, ttl);
          const renewedOut = renewed?.result?.output ?? {};
          if (!renewedOut.renewed) {
            throw new Error(
              "fiducia: retry-discovered semaphore permit lost fenced authority during renewal",
            );
          }
          acquired = true;
          return this.semaphoreHandle(
            key,
            holder,
            token,
            renewedOut.lease_expires_ms,
            ttl,
          );
        }
        if (retryOut.reason === "limit_mismatch") {
          throw new Error(
            `fiducia: semaphore limit mismatch (requested ${limit}, configured ${String(retryOut.limit)})`,
          );
        }
      }
      throw new LockTimeoutError([key], maxWait);
    } finally {
      if (!acquired) await this.cancelSemaphoreWait(key, holder, requestId);
    }
  }

  private semaphoreHandle(
    key: string,
    holder: string,
    fencingToken: number,
    leaseExpiresMs?: number,
    defaultTtlMs: number = DEFAULT_TTL,
  ): SemaphoreHandle {
    let currentTtlMs = defaultTtlMs;
    const handle: SemaphoreHandle = {
      key,
      holder,
      fencingToken,
      leaseExpiresMs,
      renew: async (ttlMs = currentTtlMs) => {
        const response = await this.semaphoreRenew(key, holder, fencingToken, ttlMs);
        const output = response?.result?.output ?? {};
        if (!output.renewed) throw new Error("fiducia: semaphore renewal lost fenced authority");
        handle.leaseExpiresMs = output.lease_expires_ms;
        currentTtlMs = ttlMs;
        return response;
      },
      unlock: () => this.semaphoreRelease(key, { holder, fencingToken }),
      release: () => this.semaphoreRelease(key, { holder, fencingToken }),
    };
    return handle;
  }

  private async cancelSemaphoreWait(key: string, holder: string, requestId: string): Promise<void> {
    const response = await this.semaphoreCancel(key, holder, { requestId });
    const output = response?.result?.output ?? {};
    if (output.acquired && output.fencing_token != null) {
      const released = await this.semaphoreRelease(key, {
        holder,
        fencingToken: fencingToken(output, "raced semaphore permit"),
      });
      if (released?.result?.output?.released === false) {
        throw new Error("fiducia: raced semaphore permit could not be released safely");
      }
      return;
    }
    if (output.cancelled === true) return;
    throw new Error(
      `fiducia: semaphore cancellation did not establish safety (${String(output.reason ?? "invalid_response")})`,
    );
  }
}

function normalize(keyOrKeys: string | string[]): string[] {
  return Array.isArray(keyOrKeys) ? keyOrKeys : [keyOrKeys];
}
